src/batch.js

```diff
@@ -1,28 +1,221 @@
-  maxBatchSize: 20
+  maxBatchSize: 20,
+  handler: undefined
 }
 
 module.exports = {
   batchHandler
 }
 
-async function batchHandler({ routes, config, extractPathParameters }, event, context) {
+async function batchHandler({ route, config }, event, context) {
   config = { ...batchDefaultConfig, ...config }
+  const body = event.body
 
   // Validate
-  // validate json body has been parsed
-  // validate body.requests has array of requests
-  // validate each body.requests[]
+  validateBatchRequest(body, config.maxBatchSize)
 
   // build graph of requests
+  const orderedRequests = tSort(body.requests)
+  const { asyncRequests, syncRequests } = splitAsyncRequests(orderedRequests)
 
   // execute requests
+  const results = await Promise.all([
+    (async () => {
+      const syncResults = []
+      for (let req of syncRequests) {
+        syncResults.push(await executeRequest(config, route, event, req))
+      }
+      return syncResults
+    })(),
+    ...asyncRequests.map(executeRequest.bind(null, config, route, event))
+  ])
 
   // return response
+  return {
+    statusCode: 200,
+    body:{ responses: results.map((result, index) => {
+      if (index === 0){
+        return result.map((result, index) => {
+          return {
+            id: syncRequests[index].id,
+            status: result.statusCode,
+            body: result.body ? JSON.parse(result.body) : undefined,
+            headers: result.headers
+          }
+        })
+      }
+      return {
+        id: asyncRequests[index - 1].id,
+        status: result.statusCode,
+        body: result.body ? JSON.parse(result.body) : undefined,
+        headers: result.headers
+      }
+    })},
+    headers:{}
+  }
 }
 
-function validateBatchRequest(body) {}
+async function executeRequest(config, route, event, request) {
+  //Parse url into pathParameters and multiValueQueryStringParameters formats
+  const urlAndQuerystring = request.url.split('?')
+  let queryStringParameters = {}
+  if (urlAndQuerystring.length > 1) {
+    const queryStringParams = urlAndQuerystring[1]
+    const keys = queryStringParams.split('&')
+    const queryStrings = keys.map((k) => k.split('='))
+
+    for (let queryString of queryStrings) {
+      if (queryStringParameters[queryString[0]]) {
+        queryStringParameters[queryString[0]].push(queryString[1])
+      } else {
+        queryStringParameters[queryString[0]] = [queryString[1]]
+      }
+    }
+  }
+
+  //Compose event to appear as native AWS event
+  let childEvent = { 
+    ...event,
+    httpMethod: request.method,
+    headers: {...normalizeRequestHeaders(request.headers), authorization: event.headers.authorization},
+    body: request.body,
+    path: request.url,
+    pathParameters: { 
+      proxy: urlAndQuerystring[0].split('/').splice(2).join('/') 
+    },
+    multiValueQueryStringParameters: queryStringParameters,
+    _json: true
+  }
+
+  //Allow consumers to define a custom handler 
+  if (config.handler){
```

Comment:
Can you explain whats going on here? I'm confused on what it means for a custom handler on a batch config

Reply:
Yeah I put some details in a strange place up above, but the situation I'm specifically trying to solve is, what if we want/need to run some initialization logic on every event, but would prefer to do so before routing occurs.

In our specific case we have logging and object initialization logic we always accomplish before we want to route. So we really need some way to hook into the logic before routing occurs for each event. 'handler' might not be a great name though, it does act as other handles in that all calls will be routed through it, but there an assumption that it eventually routes the child request (either with the lambda-router.route, or some internal logic).

Here's the other explanation I gave:
handler allows a consumer to bypass the flow:
event -> batchRoute -> hander+handler+handler

Instead allowing them to customize the behavior of each called handler, or provide a way to run each though the initial event flow. For example:
eventInit -> batchRoute -> eventInit -> lambda-route -> handler
or
eventInit -> batchRoute -> customAuthorizer -> customRoute -> handler

Reply:
I don't think this is necessary. If the app wants to run logic before a handler then it should provide a wrapped handler that does that.